src/comlink.ts

```diff
@@ -475,20 +478,28 @@ function createProxy<T>(
   target: object = function () {}
 ): Remote<T> {
   let isProxyReleased = false;
-  const propProxyCache : Map<(string | symbol), Remote<unknown>> = new Map();
+  const propProxyCache: Map<string | symbol, Remote<unknown>> = new Map();
   const proxy = new Proxy(target, {
     get(_target, prop) {
       throwIfProxyReleased(isProxyReleased);
       if (prop === releaseProxy) {
         return () => {
-          for (const subProxy of propProxyCache.values()) {
-            subProxy[releaseProxy]();
-          }
-          propProxyCache.clear();
           unregisterProxy(proxy);
-          releaseEndpoint(ep);
-          pendingListeners.clear();
           isProxyReleased = true;
+
+          return releaseEndpoint(ep, pendingListeners).finally(() => {
+            // Reject pending promises as these can't be resolved anymore.
+            for (const { reject } of pendingListeners.values()) {
+              reject(new Error("Endpoint closed"));
+            }
+            pendingListeners.clear();
+
+            for (const subProxy of propProxyCache.values()) {
+              // The endpoint has been closed at this point so we don't wait for release of all subproxies.
```

Comment:
What do you mean by the endpoint being closed? Do you mean it was "released"? Should this come before the "releaseEndpoint" call so that you _can_ wait on the sub proxies to be released rather than _void_ing them? And then release the top level endpoint proxy?